### ISSUE TITLE
Latent vectors being too "similar" with lots of 0 values

### ISSUE THREAD

**Issue by maeotaku** [dreamquark-ai/tabnet#329]
**Describe the bug**

So im not sure if this is a bug, but I wanted to use the latent vectors coming out of the encoder for other stuff. In general, I have noticed that they have tons of 0s because of a ReLU. But even if i change it for another activation, the resulting latents are very similar between each other. I wonder if this is a per design. In general, I wonder how the decoder can work with such extremely similar verctors, specially at low dimensionality.

**What is the current behavior?**
Latents are very similar between each other, not sure if they can be used or explored.

**If the current behavior is a bug, please provide the steps to reproduce.**
 Just train a pretraining model, then explore the latents from the encoder

**Expected behavior**

I would expect to see rich latent vectors, not something where most are just 0s and very few changes. 

**Screenshots**

**Other relevant information:**
poetry version:  
python version: 
Operating System: 
Additional tools:

**Additional context**

**Comment by eduardocarvp** [comment]
Hello @maeotaku ,

Could you give us a bit more insight on the data you are using?
Are you using embeddings for categorical features, default network parameters?

Not as well that there is a masking process involved during the pretraining step and that the unsupervised loss only acts on the masked values.

**Comment by Optimox** [comment]
Ok I think this is where the problems come from.

The pretraining loss used by pytorch-tabnet is normalized, a value above 1 indicates a prediction worse than predicting the mean of each feature.

Here it seems that the pretraining is doing a terrible job. Before looking at the encoder representation you should make sure that your final loss is smaller than 1.

The metric explodes when a feature has constant value across one batch. So maybe you could try a bigger batch size.
It's hard for me to help you with clear instructions on how to make your pretraining work but you should definitely try until you get a score bellow 1 otherwise the results won't make sense anyway.

I'd be happy to know if you manage to pretrain correctly and what you had to change.

Here is an example of a pretraining that actually learns something : https://www.kaggle.com/optimo/tps-apr-2021-tabnet-baseline
